### ISSUE TITLE
Multiple accesses into critical section

### ISSUE THREAD

**Issue by tomkocyan** [Darkseal/LockProvider#1]
Am I doing something wrong?

![image](https://user-images.githubusercontent.com/20599616/101205073-82e03900-366d-11eb-91ea-492f461e693b.png)

**Comment by coderdnewbie** [comment]
I think this may be a problem with GetOrAdd. This  code snippet and article explains in more detail:

https://dotnetfiddle.net/Mobile?id=7w5bS5#code-editor

https://andrewlock.net/making-getoradd-on-concurrentdictionary-thread-safe-using-lazy/

**Comment by Darkseal** [comment]
Using `LazyConcurrentDictionary` might definitely be worth it in this scenario, but it doesn't seem to be the cause of the issue experienced by the OP.

The issue lied in the `Release` method, which I think I managed to fix with the latest commit (see screenshot below):

![lockprovider-testcase](https://user-images.githubusercontent.com/530429/111793609-b87bfb80-88c5-11eb-984c-30201efec48a.jpg)

I had to create a specific wrapper to keep track of the waiters so that I could `Dispose` the `SemaphoreSlim` object when it's not need anymore. I still have to properly test it, but it seems solid enough.

**UPDATE**: replaced `ConcurrentDictionary` with `LazyConcurrentDictionary` as well.
